[PXCT-1431] - Opta Edge AI Application Note (Dual Core) (#2838)
arduino/docs-content: content/hardware/07.opta/opta-family/opta/tutorials/24.anomaly-detection-application-note/content.md

***__Important note__: For accurate measurements, you should calibrate your specific ADXL335 breakout board by placing it flat on a table with the Z-axis pointing up and measuring the actual voltage outputs. The X and Y axes should read approximately the same voltage (this is your zero-g bias point), while the Z-axis should read higher due to gravity (1g acceleration). The difference between the Z-axis voltage and the zero-g bias point gives you the sensitivity in mV/g. This calibration accounts for variations in onboard regulators and manufacturing tolerances.***

### Edge Impulse Data Formatting

The final step formats our acceleration data so it can be used with Edge Impulse data collection tools:

```arduino
// Output CSV format for Edge Impulse
Serial.print(xAccel, 4);
Serial.print(",");
Serial.print(yAccel, 4);
Serial.print(",");
Serial.println(zAccel, 4);
```

In this code:

- CSV format with four decimal places gives us the precision needed for machine learning training
- Single-line output per sample makes it easy to integrate with the Edge Impulse data forwarder

> - Comma separation follows standard CSV format that most data processing tools expect

Reviewer: ```suggestion
- Comma separation follows the standard CSV format that most data processing tools expect
```
Minor suggestion

```bash
edge-impulse-data-forwarder
```

The tool will guide you through the setup process:

1. **Login**: Enter your Edge Impulse username/email and password when prompted
2. **Select Device**: Choose the correct serial port for your Arduino (for example, `COM5`)
3. **Data Detection**: The tool will automatically detect the data frequency (100 Hz) and number of sensor axes (3)
4. **Name Axes**: When asked "What do you want to call them?", enter: `X`,`Y`,`Z`
5. **Device Name**: Give your device a name (for example, `opta`)
6. **Project Connection**: If you have multiple projects, the tool will ask which Edge Impulse project you want to connect the device to. Select your motor anomaly detection project.

![Setting up the data forwarder tool](assets/data-forwarder-configuration.png)

Once configured, the forwarder will stream data from your Opta™ board to Edge Impulse. You can verify the device connection by checking the "Devices" tab in Edge Impulse Studio. You can then start collecting training data for your machine learning model.

![Device connection verification to Edge Impulse Studio](assets/device-verification.png)

#### Data Collection Process

> With the data forwarder running, you can now collect training data for your anomaly detection model. For effective anomaly detection, you need high-quality data representing normal motor operation in different states.

Reviewer: ```suggestion
With the data forwarder running, you can now collect training data for your anomaly detection model. For effective anomaly detection, you need high-quality data that represent normal motor operation across different states.
```
Minor suggestion

![Classifier on Edge Impulse Studio](assets/classifier.png)

1. **Train Anomaly Detection Model**: Navigate to the "Anomaly detection" tab and configure K-means clustering with 32 clusters for pattern recognition. Use "Select suggested axes" to automatically choose the most relevant spectral features based on the calculated feature importance, then start training.

![Anomaly detection on Edge Impulse Studio](assets/anomaly-detection.png)

5. **Model Validation**: Test both models using the validation data to ensure the classifier accurately distinguishes between idle and nominal states, and the anomaly detector properly identifies normal operation patterns.

***__Important note__: The feature explorer shows how well your idle and nominal data separate in the feature space. Good separation means the model can clearly distinguish between different operating states. If the data points overlap significantly, you may need to collect more diverse data or adjust your sensor mounting.***

> The classification model learns to distinguish between known operating states, while the anomaly detection model creates clusters representing all normal operation patterns. This combination allows the system to both identify the current operating mode and detect unusual conditions that don't fit any normal pattern.

Reviewer: ```suggestion
The classification model learns to distinguish between known operating states, while the anomaly detection model creates clusters that represent all normal operating patterns. This combination allows the system to both identify the current operating mode and detect unusual conditions that don't fit any normal pattern.
```
Minor suggestion

- **Validation Testing**: Test both models with known normal conditions to reduce false positives and ensure reliable operation

![Model validation on Edge Impulse Studio](assets/model-validation.png)

***__Important note__: The anomaly threshold is critical for system performance. A lower threshold makes the system more sensitive and catches subtle problems but may trigger false alarms. A higher threshold reduces false alarms but might miss early-stage faults.***

### Model Deployment

After successful training and validation of both models, deploy them as an Arduino library for embedded inference:

1. **Deployment Section**: Navigate to the "Deployment" tab in Edge Impulse Studio
2. **Arduino Library**: Select "Arduino library" as the deployment target
3. **Optimization Settings**: Choose `int8` quantization for memory efficiency on the Opta™
4. **Model Analysis**:Review memory usage and inference timing estimates
5. **Download Library**: Download the generated Arduino library ZIP file
6. **Library Installation**: Install in the Arduino IDE using "Sketch > Include Library > Add .ZIP Library"

![Model deployment on Edge Impulse Studio](assets/model-deployment.png)

> The generated library includes optimized inference code for both the neural network classifier and K-means anomaly detector, specifically compiled for the Opta's Arm® Cortex®-M4 and M7 processors. This enables efficient real-time classification of operating states (idle/nominal) and detection of anomalous conditions with low computational overhead.

Reviewer: ```suggestion
The generated library includes optimized inference code for both the neural network classifier and K-means anomaly detector, specifically compiled for Opta's Arm® Cortex®-M4 and M7 processors. This enables efficient real-time classification of operating states (idle/nominal) and detection of anomalous conditions with low computational overhead.
```
Minor suggestion

## Improving the Vibration Monitor with Machine Learning

Now that we have trained our machine learning models, we can create a smart vibration monitor that automatically detects motor problems in real-time.

> The enhanced system does two things: it identifies whether the motor is running (nominal) or stopped (idle), and it alerts you when it detects unusual vibration patterns that might indicate a problem. This all happens directly on the Opta™ without needing an internet connection, leveraging the dual-core STM32H747XI microcontroller's processing capabilities for edge inference.

Reviewer: ```suggestion
The enhanced system does two things: it identifies whether the motor is running (nominal) or stopped (idle), and it alerts you when it detects unusual vibration patterns that might indicate a problem. This all happens directly on the Opta™ without needing an internet connection. The system leverages the dual-core STM32H747XI microcontroller's processing capabilities for edge inference.
```
Minor suggestion

```arduino
void triggerAnomalyAlert() {
  unsigned long currentTime = millis();

  if (currentTime - lastAlert > alertDuration) {
    // Activate relay output for external alarm
    digitalWrite(alertRelay, HIGH);

    // Visual alert pattern on the alert LED
    for (int i = 0; i < 3; i++) {
      digitalWrite(alertLED, HIGH);
      delay(150);
      digitalWrite(alertLED, LOW);
      delay(150);
    }

    // Log the event
    ei_printf("*** ANOMALY DETECTED at %lu ms - CHECK EQUIPMENT ***\n", currentTime);

    lastAlert = currentTime;
  }
}
```

When the system detects unusual vibration patterns, it activates relay output D0 (which can trigger external alarm systems such as sirens, indicator lights or SCADA notifications), flashes the onboard alert LED three times and prints a warning message. The alert system prevents spam by waiting at least 2 seconds between alerts, even if multiple anomalies are detected. During normal operation, the relay and alert LED are deactivated to ensure clean signaling.

> ***__Important note__: The Opta's relay outputs are rated for 250 VAC / 10A, making them suitable for directly driving industrial alarm devices. However, ensure that any connected alarm equipment is properly rated and that wiring complies with local electrical safety standards.***

Reviewer: ```suggestion
***__Important note__: Opta's relay outputs are rated for 250 VAC / 10A, making them suitable for directly driving industrial alarm devices. However, ensure that any connected alarm equipment is properly rated and that wiring complies with local electrical safety standards.***
```
Minor suggestion

After uploading the enhanced sketch to the Opta™, you should see the following output in the Arduino IDE's Serial Monitor during normal operation:

![Enhanced example sketch output showing real-time anomaly detection](assets/ml-monitor-output.png)

When an anomaly is detected, the alert LED will flash three times, the relay output will activate for external alarm systems, and the serial output will display the anomaly score above the configured threshold along with a warning message.

### Complete Enhanced Example Sketch

The complete intelligent motor anomaly detection sketch can be downloaded [here](assets/motor_anomaly_detection.zip).

[![ ](assets/download-button.png)](assets/motor_anomaly_detection.zip)

## Advanced: Dual-Core Anomaly Detection

> The Opta's STM32H747XI microcontroller features two independent processor cores: an Arm® Cortex®-M7 running at 480 MHz and an Arm® Cortex®-M4 running at 240 MHz. This section shows how to split the anomaly detection system across both cores for a more responsive and industrially robust architecture.

Reviewer: ```suggestion
The Opta's STM32H747XI microcontroller features two independent processor cores: an Arm® Cortex®-M7 running at 480 MHz and an Arm® Cortex®-M4 running at 240 MHz. This section shows how to split the anomaly detection system across both cores to achieve a more responsive, industrially robust architecture.
```
Minor suggestion

6. Click **Upload**. Wait for the upload to complete.
7. Open the M4 sketch (`anomaly_detection_m4.ino`).
8. Go to **Tools > Target Core** and select **M4 Co-processor**.
9. Click **Upload**. Wait for the upload to complete.
10. Open the Serial Monitor at 115200 baud to observe inference results from the M7 core.

***If the Opta™ has been previously used with the Arduino PLC IDE, run the memory partitioning sketch first. Refer to the [Opta™ Memory Partitioning for Use with the Arduino IDE](https://docs.arduino.cc/tutorials/opta/memory-partitioning) tutorial for instructions.***

### Complete Enhanced Example Sketch

The complete advanced intelligent motor anomaly detection sketch can be downloaded [here](assets/motor_anomaly_detection.zip).

[![ ](assets/download-button.png)](assets/motor_anomaly_detection.zip)

### System Integration Considerations

When deploying the intelligent anomaly detection system in industrial environments, consider the following factors based on your sensor choice:

- **Environmental Protection**: Protect the Nano R4 board and accelerometer from dust, moisture and temperature extremes using appropriate enclosures rated for the operating environment.

> - **Mounting Stability**: Ensure secure mechanical mounting of both the accelerometer sensor and the Nano R4 enclosure to prevent sensor movement that could affect measurement accuracy.

Reviewer: ```suggestion
- **Environmental Protection**: Protect Opta and accelerometer from dust, moisture and temperature extremes using appropriate enclosures rated for the operating environment.
- **Mounting Stability**: Ensure secure mechanical mounting of both the accelerometer sensor and Opta to prevent sensor movement that could affect measurement accuracy.
```
Please check this section and related details regarding board mentions :)

- **Power Management**: Implement appropriate power supply filtering and protection circuits, especially in electrically noisy industrial environments with motor drives and switching equipment.
- **Calibration Procedures**: Establish baseline measurements for each motor installation to account for mounting variations and motor-specific characteristics that may affect anomaly thresholds.
- **Maintenance Integration**: Plan integration with existing maintenance management systems through data logging interfaces or communication protocols for complete predictive maintenance programs.

## Conclusions

> This application note demonstrates how to implement motor anomaly detection using the Opta™ with either an ADXL335 accelerometer, combined with Edge Impulse machine learning platform for industrial predictive maintenance applications.

Reviewer: ```suggestion
This application note demonstrates how to implement motor anomaly detection using the Opta™ with an ADXL335 accelerometer, combined with the Edge Impulse machine learning platform, for industrial predictive maintenance applications.
```
Minor suggestion

> The solution combines the Opta's 32-bit processing power with Edge Impulse's machine learning tools to enable real-time anomaly detection directly on the embedded device. This eliminates the need for cloud connectivity and provides immediate response to potential equipment issues with inference times under 20 milliseconds.

Reviewer: ```suggestion
The solution combines Opta's 32-bit processing power with Edge Impulse's machine learning tools to enable real-time anomaly detection directly on the embedded device. This eliminates the need for cloud connectivity and provides immediate response to potential equipment issues with inference times under 20 milliseconds.
```
Minor suggestion
